CONTRIBUTING.md

```diff
@@ -0,0 +1,102 @@
+1. Rebases cleanly on the `master` branch
+1. Passes all continuous integration tests
+1. Conforms to the git commit message [format](#commit-message-format)
+1. Receives approval from another contributor
+1. Receives approval from a maintainer (distinct from the contributor review)
+
+These criteria may be satisfied in any order, but in practice your PR is unlikely to get attention from contributors until 1-3 are satisfied. Additionally, maintainer attention a scarce resource, so generally maintainers wait for a review from a non-maintainer contributor before reviewing your PR.
+
+## Code style
+The Python code in garage conforms to the [PEP8](https://www.python.org/dev/peps/pep-0008/) standard. Please read and understand it in detail.
+
+### garage-specific Python style
+These are garage-specific rules which are not part of the aforementioned style guides.
+* Python package imports should be sorted alphabetically within their PEP8 groupings. The sorting is alphabetical from left to right, ignoring case and Python keywords (i.e. `import`, `from`, `as`).
+* We prefer double-quoted strings (`"foo"`) over single-quoted strings (`'foo'`), unless there is a compelling escape or formatting reason for using single quotes
+
+### Other languages
+Non-Python files (including XML, HTML, CSS, JS, and Shell Scripts) should follow the [Google Style Guide](https://github.com/google/styleguide) for that language
+
+YAML files should use 2 spaces for indentation.
+
+### Whitespace (all languages)
+* Use Unix-style line endings
+* Trim trailing whitespace from all lines
+* All files should end in a single newline
+
+## Git
+
+### Workflow
+__garage uses a linear commit history and rebase-only merging.__
+
+This means that no merge commits appear in the project history. All pull requests, regardless of number of commits, are squashed to a single atomic commit at merge time.
+
+Do's and Don'ts avoiding accidental merge commits and other headaches:
+* *Don't* use GitHub's "Update branch" button on pull requests, no matter how tempting it seems
+* *Don't* use `git merge`
+* *Don't* use `git pull` (unless git tells you that your branch can be fast-forwarded)
+* *Don't* make commits in the `master` branch--always use a feature branch
+* *Do* fetch upstream frequently and keep your `master` branch up-to-date with upstream
+* *Do* rebase your feature branch on `master` frequently
+
+Please see [this guide](https://gist.github.com/markreid/12e7c2203916b93d23c27a263f6091a0) for a tutorial on the workflow. Note: unlike the guide, we don't use separate `develop`/`master` branches, so all PRs should be based on `master` rather than `develop`
+
+### Commit message format
+garage follows the git commit message guidelines documented [here](https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53) and [here](https://chris.beams.io/posts/git-commit/). 
+
+In short:
+* All commit messages have an informative subject line of 50 characters
```

Comment:
Could it be negotiated to 64 characters instead of 50? Sometimes a few characters make the title more clear.

Reply:
50 chars is a long-standing standard in open source software, since it's what fits comfortably on an 80 char-wide terminal with `git log --oneline`.

Even if we wanted our own, this would imply custom tooling/editor configs for highlighting and stuff. Contrast with the current standard, which is even enforced automatically by vim (if you do `git config --global core.editor=vim` you'll get some nice highlighting enforcing exactly this standard).

Reply:
FWIW, if I add automated checks, the 50 char subject will probably be treated as a warning while the 72 char body will be an error